Replace AWS_FATAL_ASSERT with explicit AWS_FATAL_PRECONDITION and AWS…

tegansiobhan opened the pull request:
…_FATAL_POSTCONDITION macros

*Issue #, if available:*

*Description of changes:*

Created new AWS_FATAL_PRECONDITION_1, AWS_FATAL_PRECONDITION_2, AWS_FATAL_POSTCONDITION_1, and AWS_FATAL_POSTCONDITION_2 macros. Replaced calls to AWS_FATAL_ERROR with these macros. 


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

angelhof commented on include/aws/common/string.h:
```diff
@@ -68,7 +68,7 @@ AWS_STATIC_IMPL const uint8_t *aws_string_bytes(const struct aws_string *str) {
  */
 AWS_STATIC_IMPL bool aws_string_eq(const struct aws_string *a, const struct aws_string *b) {
-    AWS_FATAL_ASSERT(a != NULL && b != NULL);
+    AWS_FATAL_PRECONDITION_1(a != NULL && b != NULL);
```
I would suggest using `AWS_FATAL_PRECONDITION` instead of `AWS_FATAL_PRECONDITION_1`. Similarly for postconditions.

tegansiobhan replied:
Thanks! Have made the changes!

danielsn commented:
We should consider whether we want to rename ```AWS_PRECONDITION``` to ```AWS_ASSERT_PRECONDITION``` for clarity.

feliperodri commented on include/aws/common/common.h:
```diff
@@ -115,30 +115,41 @@
 #define AWS_PRECONDITION(...) GET_MACRO(__VA_ARGS__, AWS_PRECONDITION_2, AWS_PRECONDITION_1, UNUSED)(__VA_ARGS__)
+#define AWS_FATAL_PRECONDITION(...) GET_MACRO(__VA_ARGS__, AWS_FATAL_PRECONDITION_2, AWS_FATAL_PRECONDITION_1, UNUSED)(__VA_ARGS__)
 #define AWS_POSTCONDITION(...) GET_MACRO(__VA_ARGS__, AWS_POSTCONDITION_2, AWS_POSTCONDITION_1, UNUSED)(__VA_ARGS__)
+#define AWS_FATAL_POSTCONDITION(...) GET_MACRO(__VA_ARGS__, AWS_FATAL_POSTCONDITION_2, AWS_FATAL_POSTCONDITION_1, UNUSED)(__VA_ARGS__)
```
Since you're changing these macros, could you also add `AWS_ERROR_PRECONDITION` and `AWS_ERROR_POSTCONDITION` macros that would raise error codes when the condition is violated? Something like...

```
#define AWS_ERROR_PRECONDITION(cond, err)       \
    do {                                        \
        if (cond) {                             \
            return aws_raise_error(err);        \
        }                                       \
    } while (0)
```
@danielsn @ColdenCullen what do you think?

ColdenCullen replied:
THIS IS COOL. I think it can be done in a separate PR tho, let's get just this in first.

danielsn replied:
Glad you like it.  Working on a PR now.

Also, for consistency, should I rename the current one that asserts ```AWS_ASSERT_PRECONDITION```? Or just leave it as is?

ColdenCullen replied:
I don't feel strongly either way? I don't mind the idea of all 3 being different, but I also like that ASSERT is the default. @JonathanHenson @graebm @justinboswell any thoughts?

angelhof commented on include/aws/common/array_list.inl:
```diff
@@ -86,7 +86,26 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
     bool data_is_valid = ((list->current_size == 0 && list->data == NULL) || AWS_MEM_IS_WRITABLE(list->data, list->current_size));
-    return required_size_is_valid && current_size_is_valid && data_is_valid;
+    bool item_size_is_valid = (list->item_size!=0);
+    return required_size_is_valid && current_size_is_valid && data_is_valid && item_size_is_valid;}
```
The brace should be in a new line? Did clang-format miss that?

angelhof commented on include/aws/common/array_list.inl:
```diff
@@ -86,7 +86,26 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+}
+
+AWS_STATIC_IMPL
+void aws_array_list_debug_print(const struct aws_array_list * list) {
```
Maybe this (and other debug print functions) should go in a different file debug.c or something.

tegansiobhan replied:
Yeah if there are other debug print functions we could do that, but I think that would be separate from this PR.

angelhof commented on source/priority_queue.c:
```diff
@@ -337,13 +340,19 @@ static int s_remove_node(struct aws_priority_queue *queue, void *item, size_t it
 
     aws_array_list_pop_back(&queue->container);
-    aws_array_list_pop_back(&queue->backpointers);
+
```
It seems possible to merge get_at and pop_back to happen behind one is_wiped check. 
```
    if (!aws_array_list_is_wiped(&queue->backpointers)) {
        aws_array_list_get_at(&queue->backpointers, &backpointer, swap_with);
        if (backpointer) {
            backpointer->current_index = SIZE_MAX;
         }
         aws_array_list_pop_back(&queue->backpointers);
    }
```

tegansiobhan replied:
Done and pushed.

tegansiobhan commented:
To fix broken proofs in aws_priority_queue_push_ref and aws_priority_queue_push, added the assumption that item_size must be non-zero into aws_array_list_is_valid(...). In order to account for backpointers (which can be an all zero struct), added a new function aws_array_list_is_wiped(...). When priority queue functions are called on a backpointer, this criteria is now checked first. @dsn

angelhof commented on include/aws/common/array_list.inl:
```diff
@@ -86,9 +91,34 @@ bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+bool aws_array_list_is_wiped(const struct aws_array_list *AWS_RESTRICT list) {
+    if(!list) {
+        return false;
+    }
```
Should we return `false` if list is NULL? Or should this be checked in an assert?

angelhof replied:
@danielsn what do you think?

tegansiobhan commented:
@JonathanHenson The build fail for aws-c-common-downstream is due to a bug in aws-c-io that is unrelated to this PR.

tegansiobhan commented:
@danielsn @ColdenCullen Do you think this PR can be merged now?

danielsn commented:
Downstream failure is 
```
/tmp/build/downstream/aws-c-http/tests/test_connection.c:24:10: fatal error: 'aws/io/log_writer.h' file not found 
#include <aws/io/log_writer.h> 
         ^~~~~~~~~~~~~~~~~~~~~ 
1 error generated. 
```
